rust/noosphere-p2p/src/dht/node.rs

```diff
@@ -0,0 +1,180 @@
+use crate::dht::{
+    channel::message_channel,
+    errors::DHTError,
+    processor::DHTProcessor,
+    types::{DHTMessageClient, DHTNetworkInfo, DHTRequest, DHTResponse, DHTStatus},
+    DHTConfig,
+};
+use std::time::Duration;
+use tokio;
+
+macro_rules! ensure_response {
+    ($response:expr, $matcher:pat => $statement:expr) => {
+        match $response {
+            $matcher => $statement,
+            _ => Err(DHTError::Error("Unexpected".into())),
+        }
+    };
+}
+
+/// Represents a DHT node running in a network thread, providing
+/// async methods for operating the node. Use [DHTNode::spawn] to
+/// spawn a new DHT node, returning the corresponding [DHTNode].
+pub struct DHTNode {
+    config: DHTConfig,
+    state: DHTStatus,
+    client: Option<DHTMessageClient>,
+    thread_handle: Option<tokio::task::JoinHandle<Result<(), DHTError>>>,
+}
+
+impl DHTNode {
+    fn new(config: DHTConfig) -> Result<Self, DHTError> {
+        let (client, processor) = message_channel::<DHTRequest, DHTResponse, DHTError>();
+        let thread_handle = DHTProcessor::spawn(config.clone(), processor)?;
+
+        Ok(DHTNode {
+            config,
+            state: DHTStatus::Active,
+            client: Some(client),
+            thread_handle: Some(thread_handle),
+        })
+    }
+
+    /// Spawns a networking thread with a DHT node, returning the
+    /// corresponding [DHTNode].
+    pub fn spawn(config: DHTConfig) -> Result<DHTNode, DHTError> {
```

Comment:
This could probably just be removed in lieu of `new`, but did not want to trivialize starting up a listener for a network